crates/matrix-sdk/src/room/common.rs

```diff
@@ -879,3 +919,102 @@ impl<'a> MessagesOptions<'a> {
         })
     }
 }
+
```

Comment:
I think this is more complex than it needs to be. I think this is because we're adding an effect to those pure data types so that we can execute the effect in an uniform way.

Creating an enum over those data types and letting the `decrypt()` method take `impl Into<EventEnum>` would achieve the same, while the data type would remain pure.

```rust
pub enum EventEnum<'a> {
    Foo(&'a SyncRoomEvent),
    Bar(&'a RoomEvent),
    ...
}
```